### ISSUE TITLE
Lets Encrypt  unable to authorize.

### ISSUE THREAD

**Issue by mrubinsk** [sourcegraph/issues#99]
I was running with tls.letsencrypt set to "auto" for quite some time, and a quite a few versions of Sourcegraph.  I can't pin point when this started to not work since my certificate just expired.  The sourcegraph log shows:

http: TLS handshake error from 74.8.75.11:62801: acme/autocert: unable to authorize "xxx.xxx"; tried ["tls-sni-02" "tls-sni-01" "http-01"]
21:32:46       frontend | http: TLS handshake error from 74.8.75.11:62802: acme/autocert: missing certificate

and I can now no longer access SG over https at all.

I'm wondering if it's related to:
https://github.com/golang/go/issues/21890

**Comment by keegancsmith** [comment]
Thanks for the report. `http-01` should still be working. I will try reproduce and fix.

**Comment by keegancsmith** [comment]
Hi @mrubinsk. I just setup a fresh instance and LetsEncrypt worked for me. Do you mind sharing your Sourcegraph configuration? Please redact any secrets in it.

**Comment by mrubinsk** [comment]
Sure. 

I currently have a http endpoint in my config so I can actually access it. I'd normally use the following - I left the tls.letsencrypt in the config to hightlight it's set to auto, even though it's the default value. 

The site was working fine, but in the morning yesterday the certificate was expired and it was not auto updating. I cycled the docker container thinking it might help.

Here is the command that I used to start the container:

docker run  -d --restart always --publish 80:7080 --publish 443:7443  --volume $HOME/.sourcegraph/config:/etc/sourcegraph  --volume $HOME/.sourcegraph/data:/var/opt/sourcegraph  -v /var/run/docker.sock:/var/run/docker.sock sourcegraph/server:2.9.1


{
  "appURL": "https://sourcegraph.xxxxxxxx.xxx",
  "auth.allowSignup": false,
  "auth.providers": [
    { "type": "builtin" }
  ],
  "disablePublicRepoRedirects": true,
  "update.channel": "release",
  "maxReposToSearch":500,
  "repoListUpdateInterval": 15,
  "tls.letsencrypt": "auto",
"langservers": [
  {
    "address": "tcp://typescript:2088",
    "disabled": true,
    "language": "javascript"
  },
  {
    "address": "tcp://java:2088",
    "disabled": true,
    "language": "java"
  },
  {
    "address": "tcp://php:2088",
    "disabled": true,
    "language": "php"
  },
  {
    "address": "tcp://python:2087",
    "disabled": true,
    "language": "python"
  },
  {
    "address": "tcp://typescript:2088",
    "disabled": true,
    "language": "typescript"
  },
  {
    "address": "tcp://cs:8080",
    "disabled": true,
    "language": "cs"
  },
  {
    "address": "tcp://html:8080",
    "disabled": true,
    "language": "html"
  },
  {
    "address": "tcp://css:8080",
    "disabled": true,
    "language": "css"
  }
],
"repos.list":[
.
.
.
.]
}

**Comment by geekflyer** [comment]
I'm also noticing some strange behaviour with respect to SSL after upgrading from 2.8.9 to 2.9.1.
First of all after upgrading I could not reach sourcegraph anymore.
I restarted sourcegraph a couple of times and also saw once in a while the following error (this error didn't come up consistently though):
```
22:34:24       frontend | http: TLS handshake error from 34.201.15.21:51898: acme/autocert: missing server name
22:34:24       frontend | http: TLS handshake error from 19.205.41.121:48270: acme/autocert: missing server name
22:34:28       frontend | http2: server: error reading preface from client 92.74.118.173:7907: read tcp 172.17.0.2:7443->92.74.118.173:7907: read: connection reset by peer
```

What surprisingly fixed for me the situation was setting  `httpToHttpsRedirect` from `on` to `off` (I used to have set it to `on` as suggested in the SSL/TLS docs).
Note that even after setting `httpToHttpsRedirect` to `off` it still seems to perform a redirect every time I access sourcegraph. Also, as a safety measure after setting `httpToHttpsRedirect` to off I closed port 80 in my VM firewall (now only 443 remains open).

Here's my redacted sourcegraph config:

```json
{
  "appURL": "https://redacted.acme.com",
  "maxReposToSearch": 50,
  "update.channel": "release",
  "httpToHttpsRedirect": "off",
  "gitlab": [
    {
      "url": "https://gitlab.com",
      "token": "__redacted__"
    }
  ],
  "auth.providers": [
    {
      "type": "openidconnect",
      "issuer": "https://accounts.google.com", // All G Suite domains use this issuer URI.
      "clientID": "__redacted__",
      "clientSecret": "__redacted__",
      "requireEmailDomain": "redacted.com"
    }
  ],
  "langservers": [
    {
      "address": "tcp://typescript:2088",
      "language": "javascript"
    },
    {
      "address": "tcp://python:2087",
      "language": "python"
    },
    {
      "address": "tcp://go:4389",
      "language": "go"
    },
    {
      "address": "tcp://java:2088",
      "language": "java"
    },
    {
      "address": "tcp://typescript:2088",
      "language": "typescript"
    }
  ]
}
```

**Comment by mrubinsk** [comment]
@geekflyer, I had the redirect issue as well, but saw it was probably due to https://github.com/sourcegraph/issues/issues/98. Which was apparently just fixed with the latest release.  Still can't authorize LE using the latest, though.

**Comment by beyang** [comment]
@geekflyer after you upgrade to 2.9.2, are you able to turn `httpToHttpsRedirect` back to `on`?

@mrubinsk I have also tried to reproduce this issue with 2.9.2, but have not succeeded. One possibility is that port 80 is no longer publicly accessible on your host machine (sometimes folks remove ingress to port 80 once HTTPS is set up). The LetsEncrypt `http-01` challenge requires port 80 to be publicly accessible. A quick way to test if it is available is to `curl sourcegraph.xxxxxxxx.xxx` and see if it returns something or times out.

**Comment by geekflyer** [comment]
yes @beyang upgrading to 2.9.2 fixed the httpToHttpsRedirect behaviour. Thanks.

**Comment by keegancsmith** [comment]
No worries! I'm filing a follow-up issue to make this easier to diagnose, since this sounds like a common problem that may occur.
